### ISSUE TITLE
storage_proxy::query_partition_key_range_concurrent(): use more aggressive range merging

### ISSUE THREAD

**Issue by denesb** [scylladb/scylladb#6599]
Ranges scans work on vnodes resolution, scanning one vnodes in a single step. When scanning an empty table, the range scan algorithm uses exponentially increasing concurrency. In each iteration it will read concurrently 2 X more vnodes than in the iteration before, until it has enough data to fill the page. Often, especially with higher concurrency, many of the vnodes selected for the current iteration will belong to the same replica group. We do consider vnodes for merging but only merge neighbouring ranges, whose replica groups have a >= CL intersection.

Lets consider a cluster of 3 nodes: {A, B, C} and a table with RF=3, that is nearly empty. This cluster has 3 * 256 -> 768 vnodes. When scanning this table (CL=QUORUM), the algorithm will use concurrencies: 1, 2, 4, 8, 16, .... This is a queue: 2^0 + 2^1 + 2^2 + ... 2^n = 2^(n+1) - 1; The largest n such that this sum is < 768 -> 8. At step 8 (and step 9 too), we will be reading 256 ranges in parallel. We will be able to merge all neighbouring ranges as all data is on all nodes. However, with a good chance, we will still send many parallel reads to the same node. This gets worse if RF > num_nodes, in this case we will be able to merge even less ranges, while the number of vnodes will be higher (with more nodes).

We can use a more aggressive range merging strategy. We can merge all participating vnodes, into a single read, *iff* the intersection of their replica group >= CL, even if the vnodes are *not* neighbours (have holes in them). Of course if we do this naively, creating a single range which is the union of all merged vnodes, we will need to do per partition merging when building the combined query result of all sub-reads. This can be expensive, but can be avoided if we instead send a list of ranges to the replica, which would use a multi-range reader, with a special result builder, which inserts separators into the built results, marking where each per-vnode result starts and ends. The result merger on the coordinator then would build a heap of these results and take one vnode at a time and add it to the final result. This is pretty much how it happens right now, except the per-vnode results are in a vector and the merger just iterates over it. In the above example, this would cap the number of parallel read operations to 3 (edit: actually I think just 1), instead of > 200.

The motivation for coming up with this suggestion is https://github.com/scylladb/scylla/issues/6548, where I observed that a range scan on an empty table causes timeouts on the replicas, as the parallel read operations on the same replica compete with each other for the read concurrency semaphore, stepping on each other's toes, leading to timeouts on a slow disk/busy node.

**Comment by denesb** [comment]
I should also note that the reason I think we could do this is that with the current replica-local range scan algorithm, scanning the same table with a high concurrency (dividing the work between many scans) has no advantage over just one scan with the entire range.

**Comment by avikivity** [comment]
We should avoid optimizations for empty tables, since these are rare. As long as we give somewhat reasonable service, that should be enough. Moreover, this is an optimization for small cluster sizes, and becomes less effective as the cluster size increases.

#6548 is debug mode, yes? this makes it an even more specialized optimization.

**Comment by denesb** [comment]
On Sun, Jun 7, 2020 at 7:07 PM Avi Kivity <notifications@github.com> wrote:

> We should avoid optimizations for empty tables, since these are rare. As
> long as we give somewhat reasonable service, that should be enough.
> Moreover, this is an optimization for small cluster sizes, and becomes less
> effective as the cluster size increases.
>

Actually this gets worse for larger clusters, as more nodes mean more
vnodes. I think this is precisely why we see this with #6548: it launches 6
nodes instead of the usual 3 or less used in most dtests. I was expecting
the tablet answer but I agree that this is a corner-case. Empty and sparse
tables should be rare, but it is very counter intuitive for users that we
do worse with them than with well populated tables. That said I don't
recall any complaints about this from the field, so maybe it is indeed
something dtest specific.

**Comment by fee-mendes** [comment]
It turns out that full scans on top of small tables do exist in the field, and their performance is not as great as one would expect.

One specific situation we saw, for a fairly small table and a 78-shards 7-node cluster, a simple `SELECT * FROM tbl LIMIT 1 BYPASS CACHE USING TIMEOUT 100s` would take ~14 seconds to complete. In that specific scenario, each shard had less than 10MiB of data on average (therefore the whole dataset could fit into memory), and the trace information showing Scylla traversing the token ring ended up as a 30MB (86K rows) file with several accesses to disk :-)

So I eventually ended up guessing the same as @denesb described: For large clusters the situation actually gets worse. 

For anyone interested at digging through the details, see `feb54065-c7f0-4d61-9e3e-872aaf980340`

**Comment by avikivity** [comment]
Note that for such small tables one should avoid BYPASS CACHE, if the tables are queried frequently.

The general solution for this problem is tablets.

**Comment by denesb** [comment]
We are moving towards tablets, I don't see us doing this improvement.

**Comment by vladzcloudius** [comment]
> We are moving towards tablets, I don't see us doing this improvement.

Wouldn't using tablets be optional, @denesb? AFAIR it was advertised it would.

**Comment by denesb** [comment]
> > We are moving towards tablets, I don't see us doing this improvement.
> 
> Wouldn't using tablets be optional, @denesb? AFAIR it was advertised it would.

It may be optional. If a user wants fast range scans, they will have to use tablets. Range scans cannot be made fast with vnodes. They can be made a bit faster in certain cases, but its just not worth putting in the effort for little situational gain, when tablets will bring overwhelming improvements.

**Comment by mykaul** [comment]
Tablets are not going to be optional when we deprecate vnodes.

**Comment by avikivity** [comment]
We should still try to send empty pages. The query latency won't decrease (it will increase), but individual page latency will decrease. This is important for cases where the table isn't empty but instead filled with tombstones.

**Comment by denesb** [comment]
> We should still try to send empty pages. The query latency won't decrease (it will increase), but individual page latency will decrease. This is important for cases where the table isn't empty but instead filled with tombstones.

I opened a separate issue for the empty pages suggestion: https://github.com/scylladb/scylladb/issues/15752.
